### ISSUE TITLE
[Enhancement]: Improve memory allocation in the provider

### ISSUE THREAD

**Issue by YakDriver** [hashicorp/terraform-provider-aws#33553]
### Description

_The focus of this issue is in making changes that result in actual improved memory use. Testing and reporting improvements are covered by #33898._

We have made some progress improving memory in the AWS provider (#32289). However, more needs to be done. Much of this work is upstream and relates to schemas.

Recent issue #31722 has demonstrated that the increasing size of the provider, combined with the recent addition of resources with a deep and complex schema has significantly increased the peak memory requirements of using the provider. This has put pressure on memory requirements in practitioner infrastructure deploy pipelines and has resulted in OOM errors when making changes to their infrastructure.

This is an umbrella issue to help track work on optimizing memory requirements.

### Potential optimizations

1. Work with upstream teams to improve memory surrounding deep and complex schemas
2. Continue to refine regular expression use generally and specifically by standardizing equivalent expressions and wrapping them in convenience functions
3. Dig deeper into [memory and CPU profiles](https://github.com/YakDriver/awsprovmemprof/) to find new places for improvement

### References

- Upsteam improvement (🎉): https://github.com/hashicorp/terraform/pull/33482
- https://github.com/google/pprof
- https://github.com/hashicorp/terraform-provider-aws/issues/31722
- https://github.com/hashicorp/terraform-provider-aws/issues/26742
- See memory use over time: https://github.com/YakDriver/awsprovmemprof
- Regex cache: https://github.com/YakDriver/regexache
- #32289
- https://tip.golang.org/doc/gc-guide

**Comment by fatbasstard** [comment]
Hi,

is the new Global schema cache (https://github.com/hashicorp/terraform/pull/33482) part of this?

**Comment by fatbasstard** [comment]
Looks like it: https://github.com/hashicorp/terraform-provider-aws/issues/31722#issuecomment-1754234295

**Comment by YakDriver** [comment]
Yes, terraform#33482 is a cross-team effort. As the AWS provider is possibly the largest provider, it tends to be where resource issues are found.
